packages/eas-cli/src/commandUtils/EasCommand.ts

```diff
@@ -13,45 +13,124 @@ import {
 import { ensureLoggedInAsync } from '../user/actions';
 import { easCliVersion } from '../utils/easCli';
 import { setVcsClient } from '../vcs';
 import GitClient from '../vcs/clients/git';
 
-export default abstract class EasCommand extends Command {
+export interface CommandConfiguration {
   /**
    * When user data is unavailable locally, determines if the command will
-   * force the user to log in
+   * force the user to log in. By default, all commands require authentication.
+   */
+  allowUnauthenticated?: boolean;
+
+  /**
+   * Whether a command can be run outside of an Expo project directory.
+   * By default, all commands must be run inside an Expo project directory.
    */
-  protected requiresAuthentication = true;
-  protected mustBeRunInsideProject = true;
+  canRunOutsideProject?: boolean;
+}
+
+export interface ContextOptions {
+  nonInteractive: boolean;
+}
+
+export abstract class ContextField<T> {
+  abstract getValueAsync(options: ContextOptions): Promise<T>;
+}
+
+export class ProjectIdContextField extends ContextField<string> {
+  async getValueAsync({ nonInteractive }: ContextOptions): Promise<string> {
+    const projectDir = await findProjectRootAsync();
+    const exp = getExpoConfig(projectDir);
+    return await getProjectIdAsync(exp, { nonInteractive });
+  }
+}
+
+export class ActorContextField extends ContextField<Actor> {
+  async getValueAsync({ nonInteractive }: ContextOptions): Promise<Actor> {
+    return await ensureLoggedInAsync({ nonInteractive });
+  }
+}
+
+export const EASCommandProjectIdContext = {
+  projectId: new ProjectIdContextField(),
+};
+
+export const EASCommandLoggedInContext = {
+  actor: new ActorContextField(),
+};
+
+type ContextInput<
+  T extends {
+    [name: string]: any;
+  } = object
+> = {
+  [P in keyof T]: ContextField<T[P]>;
+};
+
+type ContextOutput<
+  T extends {
+    [name: string]: any;
+  } = object
+> = {
+  [P in keyof T]: T[P];
+};
+
+export default abstract class EasCommand extends Command {
+  static contextDefinition: ContextInput = {};
+
+  protected commandConfiguration: CommandConfiguration = {};
+
+  protected async getContextAsync<
+    C extends {
+      [name: string]: any;
+    } = object
+  >(
+    commandClass: { contextDefinition: ContextInput<C> },
+    { nonInteractive }: { nonInteractive: boolean }
+  ): Promise<ContextOutput<C>> {
+    const contextDefinition = commandClass.contextDefinition;
+
+    const contextValuePairs = await Promise.all(
+      Object.keys(contextDefinition).map(async contextKey => {
+        return [contextKey, await contextDefinition[contextKey].getValueAsync({ nonInteractive })];
+      })
+    );
+
+    return Object.fromEntries(contextValuePairs);
+  }
 
   protected abstract runAsync(): Promise<any>;
 
   // eslint-disable-next-line async-protect/async-suffix
   async run(): Promise<any> {
     await initAnalyticsAsync();
 
-    if (this.mustBeRunInsideProject) {
+    if (!this.commandConfiguration.canRunOutsideProject) {
```

Comment:
if the context is supposed to be the only source of truth, then we should also handle a case where the command does not have to be run inside the project, but if it is, then it can use that information.

e.g. `eas credentials` should work like that, outside of the project it is managing account credentials, and if it's in the project it can be more specific. Running outside the projects seems to be already broken on main, so it's not related to this PR, but in a case like that should be supported

Reply:
Yep, later on in this stack I introduce a OptionalProjectIdContext, which addresses this exact case.